File: 2022_04_23_advent_of_code_in_csharp/AdventOfCode/AdventOfCode2021/Day16.cs
// #pragma warning disable CS8509

using static AdventOfCode.AdventOfCode2021.Option;

namespace AdventOfCode.AdventOfCode2021;

static class Day16
{
    static IEnumerable<LList<char>> Bin(int n) =>
        n switch
        {
            1 => new LList<char>[] { new('0', null), new('1', null) },
            _ => new[] { '0', '1' }.SelectMany(h => Bin(n - 1).Select(t => new LList<char>(h, t)))
        };

    static long BitsToInt64(this IEnumerable<char> bits) =>
        bits.AggregateBack((Pow2: 1L, Sum: 0L), (a, bit) => (2 * a.Pow2, a.Sum + (bit == '0' ? 0 : a.Pow2))).Sum;

    static int BitsToInt(this IEnumerable<char> bits) => (int)BitsToInt64(bits);

    static Dictionary<char, LList<char>> Mapping =
        Bin(4).ToDictionary(bits => Convert.ToString(BitsToInt(bits.ToEnumerable()), 16).ToUpper()[0], bits => bits);


    public static IEnumerable<char> LoadData(string input) => input.SelectMany(c => Mapping[c].ToEnumerable());


    record Package;
    record ValueLiteral(int Version, int TypeId, long Value) : Package;
    record Operator(int Version, int TypeId, Package[] Packages) : Package;

    public static IEnumerable<T> Share<T>(this IEnumerable<T> source)
    {
        return source.GetEnumerator().Pipe(ShareSequenceImpl);

        static IEnumerable<T> ShareSequenceImpl(IEnumerator<T> e)
        {
            while (e.MoveNext())
            {
                yield return e.Current;
            }
        }
    }

    static long ReadValue(IEnumerable<char> reader)
    {
        return ReadValueBits(reader).Pipe(BitsToInt64);

        static IEnumerable<char> ReadValueBits(IEnumerable<char> reader)
        {
            while (true)
            {
                var first = reader.First();
                foreach (var bit in reader.Take(4))
                {
                    yield return bit;
                }
                if (first == '0')
                {
                    yield break;
                }
            }
        }
    }

    static Package[] ReadOperator(IEnumerable<char> reader) =>
        reader.First() switch
        {
            '0' => reader.Take(15).BitsToInt()
                .Pipe(bitsCount => reader.Take(bitsCount).Share())
                .Pipe(newReader => ReadManyPackage(newReader).ToArray()),
            _ => reader.Take(11).BitsToInt()
                .Pipe(packageCount => ReadManyPackage(reader).Take(packageCount).ToArray())
        };

    static Option<Package> TryReadPackage(IEnumerable<char> reader) =>
        reader.Take(3).ToList() switch
        {
            { Count: 3 } first3Bits =>
                (first3Bits.BitsToInt(), reader.Take(3).BitsToInt()) switch
                {
                    var (version, typeId) => Some<Package>(typeId == 4 ?
                        new ValueLiteral(version, typeId, ReadValue(reader)) :
                        new Operator(version, typeId, ReadOperator(reader)))
                },
            _ => None<Package>()
        };

    static IEnumerable<Package> ReadManyPackage(IEnumerable<char> reader)
    {
        while (true)
        {
            if (TryReadPackage(reader) is Some<Package>(var package))
            {
                yield return package;
            }
            else
            {
                yield break;
            }
        }
    }

    static long SumVersions(Package package) =>
        package switch
        {
            ValueLiteral(var version, _, _) => version,
            Operator(var version, _, var packages) => version + packages.Sum(SumVersions),
            _ => throw new NonExhaustivePatternMatchingException()
        };

    static long Calculate(Package package) =>
        package switch
        {
            ValueLiteral(_, _, var value) => value,
            Operator(_, var typeId, var packages) =>
                packages.Select(Calculate).Pipe(values =>
                    typeId switch
                    {
                        0 => values.Sum(),
                        1 => values.Aggregate(1L, (a, b) => a * b),
                        2 => values.Aggregate(long.MaxValue, Math.Min),
                        3 => values.Aggregate(long.MinValue, Math.Max),
                        5 => values.Pairwise().First() is var (a, b) && a > b ? 1L : 0L,
                        6 => values.Pairwise().First() is var (a, b) && a < b ? 1L : 0L,
                        7 => values.Pairwise().First() is var (a, b) && a == b ? 1L : 0L,
                        _ => throw new Exception($"Unknown package typeId={typeId}")
                    }),
            _ => throw new NonExhaustivePatternMatchingException()
        };


    static string Puzzle(string input, Func<Package, long> calculateResult)
    {
        var data = LoadData(input);
        var reader = data.Share();
        var result = TryReadPackage(reader).Pipe(Option.Get).Pipe(calculateResult);
        return result.ToString();
    }

    public static string Puzzle1(string input) => Puzzle(input, SumVersions);
    public static string Puzzle2(string input) => Puzzle(input, Calculate);




    #region statement-based code

    static Package[] ReadOperator_(IEnumerable<char> reader)
    {
        var first = reader.First();
        if (first == '0')
        {
            var bitsCount = reader.Take(15).BitsToInt();
            var newReader = reader.Take(bitsCount).Share();
            return ReadManyPackage(newReader).ToArray();
        }
        else
        {
            var packageCount = reader.Take(11).BitsToInt();
            return ReadManyPackage(reader).Take(packageCount).ToArray();
        }
    }

    static Option<Package> TryReadPackage_(IEnumerable<char> reader)
    {
        var first3Bits = reader.Take(3).ToList();
        if (first3Bits is { Count: 3 })
        {
            var version = first3Bits.BitsToInt();
            var typeId = reader.Take(3).BitsToInt();
            return Some<Package>(typeId == 4 ?
                new ValueLiteral(version, typeId, ReadValue(reader)) :
                new Operator(version, typeId, ReadOperator(reader)));
        }
        else
        {
            return None<Package>();
        }
    }

    #endregion   
}

//#pragma warning restore CS8509

File: 2021_12_24_advent_of_code_in_fsharp/AdventOfCode/AdventOfCode2021/Day16.fs
module AdventOfCode2021.Day16

open System

let rec bin n =
    if n = 1 then
        [ [ '0' ]; [ '1' ] ] :> seq<_>
    else
        [ '0'; '1' ] |> Seq.collect (fun h -> bin (n - 1) |> Seq.map (fun t -> h :: t))

let bitsToInt64 bits = Seq.foldBack (fun bit (p, r) -> (2L * p, (if bit = '0' then r else p + r))) bits (1L, 0L) |> snd

let bitsToInt bits = bits |> bitsToInt64 |> int

let mapping = bin 4 |> Seq.map (fun bits -> Convert.ToString(bitsToInt bits, 16).[0] |> Char.ToUpper, bits) |> Map

let loadData (input: string) = input |> Seq.collect (fun c -> mapping.[c])

// let loadData' (input: string) =
//     input |> Seq.collect (fun c -> Convert.ToString(Convert.ToInt32(c |> string, 16), 2).PadLeft(4, '0'))
// loadData "D2FE28" |> Seq.toArray |> String


type Package =
    | ValueLiteral of version: int * typeId: int * value: int64
    | Operator of version: int * typeId: int * packages: Package list

let shareSequence (source: seq<char>) =
    let enumerator = source.GetEnumerator()
    seq {
        while enumerator.MoveNext() do
            enumerator.Current
    }

let readValue reader =
    seq {
        let mutable finish = false
        while not finish do
            let bit = reader |> Seq.head
            yield! reader |> Seq.take 4
            finish <- bit = '0'
    }
    |> bitsToInt64


let rec readOperator reader =
    let lengthTypeId = reader |> Seq.head
    if lengthTypeId = '0' then
        let bitsCount = reader |> Seq.take 15 |> bitsToInt
        let newReader = reader |> Seq.take bitsCount |> shareSequence // shareSequence !!!
        newReader |> readManyPackages |> Seq.toList
    else
        let packageCount = reader |> Seq.take 11 |> bitsToInt
        reader |> readManyPackages |> Seq.take packageCount |> Seq.toList

and tryReadPackage reader =
    let first3Bits = reader |> Seq.truncate 3 |> ResizeArray
    if first3Bits.Count = 3 then
        let version = first3Bits |> bitsToInt
        let typeId = reader |> Seq.take 3 |> bitsToInt
        if typeId = 4 then
            let value = readValue reader
            ValueLiteral(version, typeId, value) |> Some
        else
            let packages = readOperator reader
            Operator(version, typeId, packages) |> Some
    else
        None

and readManyPackages reader =
    seq {
        match (tryReadPackage reader) with
        | Some package ->
            yield package
            yield! readManyPackages reader
        | None -> ()
    }

let rec sumVersions packages =
    match packages with
    | ValueLiteral (version, _, _) -> version
    | Operator (version, _, packages) -> version + (packages |> Seq.sumBy sumVersions)




let rec calculate package =
    match package with
    | ValueLiteral (_, _, value) -> value
    | Operator (_, typeId, packages) ->
        let values = packages |> Seq.map calculate
        match typeId with
        | 0 -> values |> Seq.sum
        | 1 -> values |> Seq.fold (*) 1L
        | 2 -> values |> Seq.fold min Int64.MaxValue
        | 3 -> values |> Seq.fold max Int64.MinValue
        | 5 -> values |> Seq.pairwise |> Seq.head ||> (>) |> (fun g -> if g then 1L else 0L)
        | 6 -> values |> Seq.pairwise |> Seq.head ||> (<) |> (fun l -> if l then 1L else 0L)
        | 7 -> values |> Seq.pairwise |> Seq.head ||> (=) |> (fun e -> if e then 1L else 0L)
        | _ -> failwith $"Unknown package typeId={typeId}"


let puzzle (input: string) calculateResult =
    let data = loadData input
    let reader = data |> shareSequence
    let result = reader |> tryReadPackage |> Option.get |> calculateResult
    result |> string

let puzzle1 (input: string) = puzzle input sumVersions
let puzzle2 (input: string) = puzzle input calculate
